tests/test_unit/test_annotations/test_io.py

```diff
@@ -0,0 +1,524 @@
+    # (this should raise an error for COCO files)
+    with expected_exception as excinfo:
+        df = _df_bboxes_from_single_specific_file(
+            file_path=annotations_test_data[input_file],
+            validator=validator,
+            get_rows_from_file=row_function,
+        )
+
+    # If no error expected, check that the dataframe has empty categories
+    if not excinfo:
+        assert all(df.loc[:, "category"] == "")
+        assert all(df.loc[:, "supercategory"] == "")
+
+
+@pytest.mark.parametrize(
+    "input_file, expected_n_annotations",
+    [
+        ("VIA_JSON_sample_1.json", 4440),
+        ("VIA_JSON_sample_2.json", 3977),
+        ("small_bboxes_VIA.json", 3),
+        ("small_bboxes_duplicates_VIA.json", 4),  # contains duplicates
+    ],
+)
+def test_df_rows_from_valid_VIA_file(
+    input_file: str,
+    expected_n_annotations: int,
+    annotations_test_data: dict,
+):
+    """Test the extraction of rows from a valid VIA file."""
+    rows = _df_rows_from_valid_VIA_file(
+        file_path=annotations_test_data[input_file]
+    )
+
+    # Check number of rows
+    assert len(rows) == expected_n_annotations
+
+    # Check each row contains required column data
+    # Note that "image_width" and "image_height" are not exported to the
+    # VIA file
+    for row in rows:
+        assert all(
+            key in row
+            for key in [STANDARD_BBOXES_DF_INDEX] + STANDARD_BBOXES_DF_COLUMNS
+            if key not in ["image_width", "image_height"]
+        )
+
+
+@pytest.mark.parametrize(
+    "input_file, expected_n_annotations",
+    [
+        ("COCO_JSON_sample_1.json", 4344),
+        ("COCO_JSON_sample_2.json", 4618),
+        ("small_bboxes_COCO.json", 3),
+        ("small_bboxes_duplicates_COCO.json", 4),  # contains duplicates
+    ],
+)
+def test_df_rows_from_valid_COCO_file(
+    input_file: str,
+    expected_n_annotations: int,
+    annotations_test_data: dict,
+):
+    """Test the extraction of rows from a valid COCO file."""
+    rows = _df_rows_from_valid_COCO_file(
+        file_path=annotations_test_data[input_file]
+    )
+
+    # Check number of rows
+    assert len(rows) == expected_n_annotations
+
+    # Check each row contains required column data
+    for row in rows:
+        assert all(
+            key in row
+            for key in [STANDARD_BBOXES_DF_INDEX] + STANDARD_BBOXES_DF_COLUMNS
+        )
+
+
+@pytest.mark.parametrize(
+    "duplicates_kwargs, expected_exception",
+    [
+        ({"ignore_index": True}, pytest.raises(ValueError)),
+        ({"inplace": True}, pytest.raises(ValueError)),
+        ({"subset": "image_id"}, does_not_raise()),
+        ({"keep": "last"}, does_not_raise()),
+    ],
+)
+@pytest.mark.parametrize(
+    "input_format, filename",
+    [
+        ("VIA", "small_bboxes_duplicates_VIA.json"),
+        ("VIA", "MULTIPLE_VIA_FILES"),
+        ("COCO", "small_bboxes_duplicates_COCO.json"),
+        ("COCO", "MULTIPLE_COCO_FILES"),
+    ],
+)
```

Comment:
Do you really need to test this behaviour for single AND multiple input files?
The error-raising behaviour you are testing for doesn't in any way depend on the number of input files (as far as I can tell).
So this second parametrisation is not really needed, or am I missing something?

Reply:
I was thinking there could be duplicates across files too and those should be removed.

But you are right that the example is not very good, because in the MULTIPLE_ fixtures there are no duplicates across files - will change that now.